cunumeric/module.py

```diff
@@ -5327,6 +5336,630 @@ def nancumsum(
     )
 
 
+def _get_non_nan_reduction_code_if_applicable(
+    dtype: npt.DTypeLike, unary_reduction_code: UnaryRedCode
+) -> UnaryRedCode:
+    """
+    Return the equivalent non-nan reduction op codes if the datatype of the
+    array doesn't belong to one of the allowed datatypes. Raise an error if the
+    datatype is disallowed, which is currently complex64 and complex128.
+    """
+
+    equivalent_non_nan_unary_red_ops: Dict[UnaryRedCode, UnaryRedCode] = {
+        UnaryRedCode.NANARGMAX: UnaryRedCode.ARGMAX,
+        UnaryRedCode.NANARGMIN: UnaryRedCode.ARGMIN,
+        UnaryRedCode.NANMAX: UnaryRedCode.MAX,
+        UnaryRedCode.NANMIN: UnaryRedCode.MIN,
+    }
+
+    assert unary_reduction_code in equivalent_non_nan_unary_red_ops
+
+    # raise error for disallowed datatypes
+    disallowed_dtypes: list[np.dtype[Any]] = [
+        np.dtype(np.complex64),
+        np.dtype(np.complex128),
+    ]
+    if dtype in disallowed_dtypes:
+        raise NotImplementedError(
+            "operation is not supported for complex64 and complex128 types"
+        )
+
+    # use non-NaN API if the datatype is not floating point type
+    allowed_dtypes: list[np.dtype[Any]] = [
+        np.dtype(np.float16),
+        np.dtype(np.float32),
+        np.dtype(np.float64),
+    ]
+    if dtype in allowed_dtypes:
+        reduction_code = unary_reduction_code
+    else:
+        reduction_code = equivalent_non_nan_unary_red_ops[unary_reduction_code]
+
+    return reduction_code
+
+
+def _handle_exceptions_and_update_out_if_needed(
+    out: ndarray,
+    op: UnaryRedCode,
+    identity: Union[int, np.floating[Any]],
+) -> None:
+    """Raise ValueError if NaN is found in a slice or in the entire
+    array for nanargmin and nanargmax. In case of nanmin and nanmax,
+    issue a RuntimeWarning.
+    """
+    supported_ops = [
```

Comment:
also move this to the global scope. in fact, I think we should move all arg reduction related utilities to a separate module.

Reply:
Yes, please, this module is already too large. FWIW in these situations I'd suggest a "private" module name with regular function name,
```python
from _foo import bar
```
rather than 
```python
from foo import _bar
```

Reply:
moved utility functions related to the current implementation to a separate file (`_unary_red_utils.py`).